cpp/locality/Voronoi.cc

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) 2010-2019 The Regents of the University of Michigan
+// This file is from the freud project, released under the BSD 3-Clause License.
+
+#include <iostream>
+#include <algorithm>
+#include <sstream>
+#include <stdexcept>
+#include <tbb/tbb.h>
+#include <tuple>
+#include <cmath>
+#include <memory>
+#include <vector>
+
+#include "Box.h"
+#include "NeighborList.h"
+#include "Voronoi.h"
+#include "NeighborQuery.h"
+
+using namespace std;
+using namespace tbb;
+
+/*! \file Voronoi.cc
+    \brief Build a cell list from a set of points.
+*/
+
+namespace freud { namespace locality {
+
+// Default constructor
+Voronoi::Voronoi()
+    {}
+
+class NeighborBondLess{
+public:
+    bool operator()(const NeighborBond &n1, NeighborBond &n2) const {
+        if (n1.index_i != n2.index_i) {
+            return n1.index_i < n2.index_i;
+        }
+        if (n1.index_j != n2.index_j) {
+            return n1.index_j < n2.index_j;
+        }
+        return n1.weight < n2.weight;
+    }
+};
+
+typedef tbb::enumerable_thread_specific< std::vector<NeighborBond> > BondVector;
+
+void add_valid_bonds(BondVector::reference local_bonds,
+    unsigned int i, unsigned int expanded_i,
+    unsigned int j, unsigned int expanded_j,
+    unsigned int N, float weight)
+{
+    // Make sure we only add bonds with real particles as the reference
+    if (i < N)
+    {
+        NeighborBond nb_ij(expanded_i, expanded_j, weight);
+        local_bonds.emplace_back(nb_ij);
+    }
+
+    if (j < N)
+    {
+        NeighborBond nb_ji(expanded_j, expanded_i, weight);
+        local_bonds.emplace_back(nb_ji);
+    }
+}
+
+void Voronoi::compute(const box::Box &box, const vec3<double>* vertices,
+    const int* ridge_points, const int* ridge_vertices, unsigned int n_ridges,
+    unsigned int N, const int* expanded_ids, const int* ridge_vertex_indices)
+    {
+        m_box = box;
+
+        // iterate over ridges in parallel
+        BondVector bonds;
+        tbb::parallel_for(tbb::blocked_range<size_t>(0, n_ridges), [&] (const tbb::blocked_range<size_t> &r) {
+             BondVector::reference local_bonds(bonds.local());
+            for (size_t ridge(r.begin()); ridge != r.end(); ++ridge) {
+                unsigned int i = ridge_points[2*ridge];
+                unsigned int j = ridge_points[2*ridge+1];
+                float weight = 0;
+
+                // Reject bonds between two image particles
+                if (i >= N && j >= N)
+                    continue;
+
+                bool exclude_ii = true;
+                // We DO allow bonds from a particle to its own image
+                if (exclude_ii && i == j)
+                    continue;
+
+                // Bonds where a ridge goes to infinity (index -1) have weight 0
+                bool weighted_bond = true;
+                vector<int> current_ridge_vertex_ids;
```

Comment:
If you change this to `vector<vec3<float> > current_ridge_vertices` and just push_back the vertex directly, it saves you some logic in the length/area calculation.